crypto/src/hash/rescue_prime/rescue_prime_optimized.rs

```diff
@@ -0,0 +1,826 @@
+use super::parameters::*;
+use super::utils::*;
+use super::Fp;
+use crate::alloc::vec::Vec;
+use core::iter;
+use lambdaworks_math::field::errors::FieldError;
+
+// Implementation of the Rescue Prime Optimized hash function.
+// https://eprint.iacr.org/2022/1577
+// https://github.com/ASDiscreteMathematics/rpo/tree/master/reference_implementation
+// It supports two security levels: 128-bit and 160-bit. Depending on the security level chosen
+// the integer parameters are set accordingly.
+
+// For the Security level (λ) of 128 bits we have:
+// Number of rounds (N): 7
+// State size (m): 12
+// Rate (r): 8
+// Capacity (c): 4
+
+// For the Security level (λ) of 160 bits we have:
+// Number of rounds (N): 7
+// State size (m): 16
+// Rate (r): 10
+// Capacity (c): 6
+
+// In the paper, the authors use a number of rounds equal to 7 as a trade-off between security and performance.
+// The number of rounds can be increased to 8 or 9 to achieve a higher level of security at the cost of performance.
+const NUM_FULL_ROUNDS: usize = 7;
+
+pub struct RescuePrimeOptimized {
+    /// State width of the hash function.
+    m: usize,
+    /// Capacity of the sponge.
+    capacity: usize,
+    /// Rate of the sponge.
+    rate: usize,
+    /// Precomputed round constants used in the permutation.
+    round_constants: &'static [Fp],
+    /// MDS matrix used in the permutation.
+    mds_matrix: Vec<Vec<Fp>>,
+    /// MDS vector used for optimizing matrix multiplication.
+    mds_vector: MdsVector,
+    /// Method used for applying the MDS matrix.
+    mds_method: MdsMethod,
+}
+
+impl Default for RescuePrimeOptimized {
+    fn default() -> Self {
+        Self::new(SecurityLevel::Sec128, MdsMethod::MatrixMultiplication).unwrap()
+    }
+}
+
+impl RescuePrimeOptimized {
+    /// Creates a new instance of `RescuePrimeOptimized` with corresponding Security level and the specified MDS method.
+    pub fn new(security_level: SecurityLevel, mds_method: MdsMethod) -> Result<Self, &'static str> {
+        let (m, capacity) = match security_level {
+            SecurityLevel::Sec128 => (12, 4),
+            SecurityLevel::Sec160 => (16, 6),
+        };
+        let rate = m - capacity;
+        let (mds_matrix, round_constants, mds_vector) = match security_level {
+            SecurityLevel::Sec128 => (
+                get_mds_matrix(SecurityLevel::Sec128),
+                get_round_constants(SecurityLevel::Sec128),
+                get_mds_vector(SecurityLevel::Sec128),
+            ),
+            SecurityLevel::Sec160 => (
+                get_mds_matrix(SecurityLevel::Sec160),
+                get_round_constants(SecurityLevel::Sec160),
+                get_mds_vector(SecurityLevel::Sec160),
+            ),
+        };
```

Comment:
This match is not needed anymore.
```suggestion
        let mds_matrix = get_mds_matrix(security_level);
        let mds_constants = get_round_constants(security_level);
        let mds_vector = get_mds_vector(security_level);
```

Reply:
You're right, thanks. Removed the match here and for size and capacity